server/src/main/java/io/crate/analyze/relations/RelationAnalyzer.java

```diff
@@ -290,30 +305,40 @@ protected AnalyzedRelation visitJoin(Join node, StatementAnalysisContext stateme
                     new SubqueryAnalyzer(this, statementContext));
-                Expression expr;
-                if (joinCriteria instanceof JoinOn) {
-                    expr = ((JoinOn) joinCriteria).getExpression();
-                } else {
-                    expr = JoinUsing.toExpression(
-                        leftRel.relationName().toQualifiedName(),
-                        rightRel.relationName().toQualifiedName(),
-                        ((JoinUsing) joinCriteria).getColumns());
-                }
-                try {
-                    joinCondition = expressionAnalyzer.convert(expr, relationContext.expressionAnalysisContext());
-                } catch (RelationUnknown e) {
-                    throw new RelationValidationException(e.getTableIdents(),
-                        String.format(Locale.ENGLISH,
-                        "missing FROM-clause entry for relation '%s'", e.getTableIdents()));
-                }
-            } else {
-                throw new UnsupportedOperationException(String.format(Locale.ENGLISH, "join criteria %s not supported",
-                    joinCriteria.getClass().getSimpleName()));
+                var expressionAnalysisContext = statementContext.currentRelationContext().expressionAnalysisContext();
+                joinCondition = expressionAnalyzer.convert(expression, expressionAnalysisContext);
+            } catch (RelationUnknown e) {
+                throw new RelationValidationException(
+                    e.getTableIdents(),
+                    String.format(Locale.ENGLISH, "missing FROM-clause entry for relation '%s'", e.getTableIdents())
+                );
             }
         }
+        var joinType = JoinType.values()[node.getType().ordinal()];
+        JoinRelation joinRelation = new JoinRelation(leftRel, rightRel, leftRel.outputs(), joinType, joinCondition);
+        JoinPair joinPair = extractJoinPair(joinRelation, statementContext.currentRelationContext().sourceNames());
+        statementContext.currentRelationContext().addJoinPair(joinPair);
+        return joinRelation;
+    }
 
-        relationContext.addJoinType(JoinType.values()[node.getType().ordinal()], joinCondition);
-        return null;
+    private static JoinPair extractJoinPair(JoinRelation joinRelation, List<RelationName> relevantRelationsInOrder) {
+        assert relevantRelationsInOrder.size() >= 2 : "sources must be added first, cannot add join pair for only 1 source";
+
+        var joinCondition = joinRelation.joinCondition();
+        if (joinCondition != null) {
+            Set<RelationName> relationsInJoinConditions = RelationNameCollector.collect(joinCondition);
+            if (relationsInJoinConditions.size() > 1) {
+                // We have join conditions with two relations or more. The join conditions such as
+                // `t1.x = t2.x` determines which relations are joined together.
```

Comment:
I think we need to ensure that Join condition has exactly 2 relations, to handle cases like 
```
SELECT * from j1 
     join j2 ON j1.x = j2.x
     join j3 on array_length(array_unique(ARRAY[j1.x, j2.x, j3.x]), 1) > 2
```

I mean we cannot build a pair from a triple and can skip it because:  
1. it's not a logical error since it doesn't affect correctness it only affects how tables will be ordered in the re-ordering optimization.  Tables, involved into such complex conditions would be considered as last ones (similar to those which are involved in `CROSS` joins)

2. It would be aligned with [getting pairs from implicit conditions](https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/execution/engine/join/JoinOperations.java#L138) (from where clause)

3. It's very synthetic/artifical example anyway, majority of conditions would fall into 2 conditions branch.

Reply:
There can be more than two relations in the join conditions, look at the pg-query example. But we only need to look at two latest relations. On a nested join each join relation will be handled separately.

Reply:
```
EXPLAIN
SELECT * from j1 
     join j2 ON j1.x = j2.x
     join j3 on array_length(array_unique(ARRAY[j1.x, j2.x, j3.x]), 1) > 2
```
yields:
```
NestedLoopJoin[INNER | (array_length(array_unique(_array(x, x, x)), 1) > 2)]
  ├ HashJoin[(x = x)]
  │  ├ Collect[ui.j1 | [x] | true]
  │  └ Collect[ui.j2 | [x] | true]
  └ Collect[ui.j3 | [x] | true]
```
This looks ok for me and it's also executable.

Reply:
My concern was that we would add pairs (t1,t2) and (t2,t3) but (t1,t3) is also a pair technically speaking and I expected it to be somehow important for ordering logic.

However, only "at least one occurrence in condition" AFAIK is important for ordering, and both t1 and t3 appear in other pairs so this looks fine indeed.

Thanks for checking and posting explain!